phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java

```diff
@@ -0,0 +1,299 @@
+                + " WHERE B >= 20) Y ON X.K=Y.K";
+        long count3 = countRowsScannedFromSql(stmt, sqlJoin);
+        assertEquals(161, count3);
+    }
+
+    @Test
+    public void testUnionAll() throws Exception {
+        Properties props = new Properties();
+        props.put(QueryServices.COLLECT_REQUEST_LEVEL_METRICS, "true");
+        // force many rpc calls
+        props.put(QueryServices.SCAN_CACHE_SIZE_ATTRIB, "10");
+        Connection conn = DriverManager.getConnection(getUrl(), props);
+        String tableName1 = generateUniqueName();
+        String tableName2 = generateUniqueName();
+        PhoenixStatement stmt = conn.createStatement().unwrap(PhoenixStatement.class);
+        stmt.executeUpdate("CREATE TABLE IF NOT EXISTS " + tableName1
+                + " (A UNSIGNED_LONG NOT NULL, Z UNSIGNED_LONG, CONSTRAINT pk PRIMARY KEY (A))");
+        stmt.executeUpdate("CREATE TABLE IF NOT EXISTS " + tableName2
+                + " (B UNSIGNED_LONG NOT NULL, Z UNSIGNED_LONG, CONSTRAINT pk PRIMARY KEY (B))");
+        for (int i = 1; i <= 100; i++) {
+            String sql1 = String.format("UPSERT INTO %s VALUES (%d, %d)", tableName1, i, i);
+            stmt.execute(sql1);
+            String sql2 = String.format("UPSERT INTO %s VALUES (%d, %d)", tableName2, i, i);
+            stmt.execute(sql2);
+        }
+
+        conn.commit();
+
+        // table1
+        long count1 = countRowsScannedFromSql(stmt,
+                "SELECT A, Z FROM " + tableName1 + " WHERE A >= 40");
+        assertEquals(61, count1);
+
+        // table2, all rows
+        long count2 = countRowsScannedFromSql(stmt,
+                "SELECT B, Z FROM " + tableName2 + " WHERE B >= 20");
+        assertEquals(81, count2);
+
+        // union all
+        String sqlUnionAll = "SELECT SUM(Z) FROM ( SELECT Z FROM " + tableName1
+                + " WHERE A >= 40 UNION ALL SELECT Z FROM " + tableName2 + " WHERE B >= 20)";
+        long count3 = countRowsScannedFromSql(stmt, sqlUnionAll);
+        assertEquals(142, count3);
+
+        // union all then group by
+        String sqlUnionAllGroupBy = "SELECT K, SUM(Z) FROM ( SELECT A AS K, Z FROM " + tableName1
+                + " WHERE A >= 40 UNION ALL SELECT B AS K, Z FROM " + tableName2
+                + " WHERE B >= 20) GROUP BY K";
+        long count4 = countRowsScannedFromSql(stmt, sqlUnionAllGroupBy);
+        assertEquals(142, count4);
+    }
+
+    private long countRowsScannedFromSql(Statement stmt, String sql) throws SQLException {
+        ResultSet rs = stmt.executeQuery(sql);
+        while (rs.next()) {
+            // loop to the end
+        }
+        return getRowsScanned(rs);
+    }
+
+    private long getRowsScanned(ResultSet rs) {
+        if (!(rs instanceof PhoenixResultSet)) {
+            return -1;
+        }
+        PhoenixResultSet phoenixResultSet = (PhoenixResultSet) rs;
+        Map<String, Map<MetricType, Long>> metrics = phoenixResultSet.getReadMetrics();
```

Comment:
Better to use `PhoenixRuntime.getRequestReadMetricInfo`

Reply:
Also, check the tableName which is the key in the map to ensure we are only extracting the relevant metrics.

Reply:
done